### DISCUSSION TITLE
The real timespan for the cheapest hours ?

### DISCUSSION THREAD

**Discussion by typxxi** [kotope/aio_energy_management#32] [Q&A]
I am currently focussing and reading only on the blog article for the aio 0.2.2 release for nordpool to avoid confusion.
I came along this description of the variables 

![image](https://github.com/user-attachments/assets/b894cbf9-fbc8-45b9-a95f-a8fa4849b74b)

In case of starting today = yes the starting hour should be after the nordpool prices have been updated

But I am wondering why cause if today means today then the nordpool prices for today have been published already yesterday or is the whole outlook / scope / horizont of the cheapest hour sensor a "longer one" like from today till this evening or till tomorrow evening if the current time is beyoned the nordpool price release time ?

So far I have only seen prices for today which makes me wonder if the sensors are searching for the cheapest price for the next 24 hours or strictly  based on the date. 

first_hour = 15 
last_hour = 23 

Does it mean a comparison starting today from 15 to 23 (as the last tarif hour of the day) aka today in the tighter sense or would it show me the cheapest hours from 15 today to tomorrow 23  or maybe revolving 24 hours from 15 today till 14 tomorrow ?

If the sun is not providing enough energy I wanna charge my batteries to avoid our peak tarifs around 18 to 20 and 6 to 9. 
So my focus must be to look beyond 23 cause usually or cheapest night prices are in the early morning depending on the wind in the north sea wind farm area cause they push the prices to 0 Cent for production (and we still have to pay 20 cent grid fees and services plus vat on top). 

The only hint for me so far is this line from the screenshot above.

`starting_today = True if the look up should be started on today. `

This sounds to me a bit like that the timespan in focus starts today 15 (after nordpool release time aka here 14) and ends with the last hour 23 today or tomorrow ?

And what if I use first_hour = 15 and last_hour=14 ? 
Would that mean a comparison for those 24 hours from 15:00 onwards only ? 

Might sound weird question but we are in the morning and my sensors are showing only hours for today due to the missing prices for tomorrow. 
That might change after 14:00 when nordpool will have released the prices for tomorrow and I might be able to answer the question if cheapest hours is limited to today by calendardate or can reach beyond midnight ?

That is important to us due to the fact that we can expect the cheapest prices during winter in the early morning, time to charge the EV and home storage in case of a bad solar power outlook or high demand in the morning. 

Maybe I have overread that in the [blog article](https://www.creatingsmarthome.com/index.php/2024/07/27/home-assistant-nord-pool-cheapest-hours-with-aio-energy-management/) i am working with cause it feels like that shows the broader picture while the github readme is shorter.

**Comment by typxxi** [comment]
Now reading along the blog I found this screenshot showing the "scope" going beyond midnight

![image](https://github.com/user-attachments/assets/9b94e31d-009c-401b-b09b-eb385fbd6733)
 
and I finally made it to the comment section where I also found a lot of usefull explanations too like that "outdated" flag which then seems to cause an update of the sensors automatically if I remember right (so from 14 to 23 would cause an update after 23:00 of the sensor with the newest prices).

**Reply by kotope** [reply]
The lifecycle in short:
- starting_today: True should be used when you want to split the 24h timespan between two days. E.g. I'm using it to make hot water between 21:00-10:00
- expiration attribute is used to know when the data is outdated e.g. lifecycle is over and next lifecycle needs to be calculated either from list_next if existing or wait for new prices. Expiration is basically same as the last_hour ending time.
- cheapest hours are 'locked in' immediately after new price data is received unless trigger_time is used. Those can't be changed until expiration is passed again.
